File: ios-native/WordWheel/Views/SpinWheel.swift
import SwiftUI

/// One sector of the lottery wheel.
struct SpinSector: Identifiable {
    let id = UUID()
    let coins: Int
    let hints: Int
    let color: Color
    let weight: Int  // probability weight

    var label: String {
        switch (coins, hints) {
        case let (c, h) where c > 0 && h > 0: return "+\(c)\n+\(h)💡"
        case let (c, _) where c > 0:           return "+\(c)"
        case let (_, h) where h > 0:           return "+\(h)💡"
        default:                                return "—"
        }
    }
}

/// Default 8-sector wheel mirroring the Android version. Casual-friendly:
/// small wins are common, rarer 50-coin / 3-hint jackpots stay exciting.
let defaultSpinSectors: [SpinSector] = [
    SpinSector(coins: 5,  hints: 0, color: Color(hex: 0xFF50A0E6), weight: 28),
    SpinSector(coins: 10, hints: 0, color: Color(hex: 0xFFFFB400), weight: 22),
    SpinSector(coins: 15, hints: 0, color: Color(hex: 0xFF50A0E6), weight: 16),
    SpinSector(coins: 0,  hints: 1, color: Color(hex: 0xFF32C850), weight: 12),
    SpinSector(coins: 20, hints: 0, color: Color(hex: 0xFFFFB400), weight: 10),
    SpinSector(coins: 25, hints: 0, color: Color(hex: 0xFF50A0E6), weight: 6),
    SpinSector(coins: 50, hints: 0, color: Color(hex: 0xFFE65050), weight: 4),
    SpinSector(coins: 0,  hints: 3, color: Color(hex: 0xFF32C850), weight: 2),
]

/// Modal lottery wheel. Tap "SPIN" → wheel spins to a weighted-random
/// sector → reward reported via `onSpinResult`. Tapping background after
/// the spin closes the dialog.
struct SpinWheelDialog: View {
    let sectors: [SpinSector]
    let onSpinResult: (SpinSector) -> Void
    let onDismiss: () -> Void

    @State private var hasSpun = false
    @State private var resultIndex = -1
    @State private var rotation: Double = 0  // degrees

    init(sectors: [SpinSector] = defaultSpinSectors,
         onSpinResult: @escaping (SpinSector) -> Void,
         onDismiss: @escaping () -> Void) {
        self.sectors = sectors
        self.onSpinResult = onSpinResult
        self.onDismiss = onDismiss
    }

    var body: some View {
        ZStack {
            Color.black.opacity(0.78)
                .ignoresSafeArea()
                .onTapGesture { if hasSpun { onDismiss() } }

            VStack(spacing: 16) {
                Text(hasSpun ? "Nice!" : "Daily Spin")
                    .font(.system(size: 26, weight: .bold))
                    .foregroundColor(.white)

                if hasSpun, let sec = sectors.indices.contains(resultIndex)
                                ? sectors[resultIndex] : nil {
                    Text(rewardSummary(sec))
                        .font(.system(size: 14))
                        .foregroundColor(GameColors.starYellow)
                } else {
                    Text("Tap below to spin — once per day")
                        .font(.system(size: 14))
                        .foregroundColor(GameColors.starYellow)
                }

                ZStack {
                    WheelShape(sectors: sectors)
                        .rotationEffect(.degrees(rotation))
                        .animation(
                            .timingCurve(0.05, 0.9, 0.4, 1.0, duration: 3.2),
                            value: rotation,
                        )
                        .frame(maxWidth: .infinity)
                        .aspectRatio(1, contentMode: .fit)

                    // Pointer (does not rotate)
                    Triangle()
                        .fill(Color(hex: 0xFFFFDC50))
                        .overlay(Triangle().stroke(Color.black.opacity(0.4), lineWidth: 3))
                        .frame(width: 36, height: 36)
                        .offset(y: -GameSpinWheelTopOffset)
                        .frame(maxHeight: .infinity, alignment: .top)
                }
                .padding(.horizontal, 4)

                Button(action: spinTapped) {
                    Text(hasSpun ? "TAP TO CLOSE" : "SPIN")
                        .font(.system(size: 18, weight: .bold))
                        .foregroundColor(.white)
                        .padding(.horizontal, 36).padding(.vertical, 14)
                        .background(RoundedRectangle(cornerRadius: 28)
                            .fill(hasSpun
                                  ? Color(hex: 0xFF555555)
                                  : Color(hex: 0xFF32C850)))
                }
                .buttonStyle(.plain)
                .disabled(hasSpun)
            }
            .padding(20)
            .frame(maxWidth: 360)
            .background(RoundedRectangle(cornerRadius: 20).fill(GameColors.completeBg))
            .padding(.horizontal, 24)
        }
    }

    private func spinTapped() {
        guard !hasSpun else { onDismiss(); return }

        let totalWeight = sectors.reduce(0) { $0 + $1.weight }
        var pick = Int.random(in: 0..<totalWeight)
        var idx = 0
        for (i, s) in sectors.enumerated() {
            if pick < s.weight { idx = i; break }
            pick -= s.weight
        }
        resultIndex = idx
        hasSpun = true

        // Wheel rotates clockwise; sector i is centred at i*sweep relative
        // to the pointer at top. 5 full turns + the negative offset lands
        // sector idx under the pointer.
        let sweep = 360.0 / Double(sectors.count)
        rotation = 360 * 5 - Double(idx) * sweep

        // Notify the caller after the animation finishes.
        DispatchQueue.main.asyncAfter(deadline: .now() + 3.3) {
            onSpinResult(sectors[idx])
        }
    }

    private func rewardSummary(_ sec: SpinSector) -> String {
        switch (sec.coins, sec.hints) {
        case let (c, h) where c > 0 && h > 0:
            return "+\(c) coins and +\(h) hints"
        case let (c, _) where c > 0:
            return "+\(c) coins"
        case let (_, h) where h > 0:
            return "+\(h) hint" + (h > 1 ? "s" : "")
        default: return ""
        }
    }
}

/// Vertical offset so the pointer sits flush against the wheel's top edge.
private let GameSpinWheelTopOffset: CGFloat = -2

/// Pie-chart-style wheel with N coloured sectors and centred labels.
private struct WheelShape: View {
    let sectors: [SpinSector]

    var body: some View {
        GeometryReader { geo in
            let radius = min(geo.size.width, geo.size.height) / 2 * 0.96
            let center = CGPoint(x: geo.size.width / 2, y: geo.size.height / 2)
            let sweep = 360.0 / Double(sectors.count)

            ZStack {
                ForEach(0..<sectors.count, id: \.self) { i in
                    let start = Double(i) * sweep - 90 - sweep / 2
                    let end = start + sweep
                    SectorShape(
                        center: center,
                        radius: radius,
                        startDeg: start,
                        endDeg: end,
                    )
                    .fill(sectors[i].color)
                    .overlay(
                        SectorShape(
                            center: center,
                            radius: radius,
                            startDeg: start,
                            endDeg: end,
                        ).stroke(Color.white.opacity(0.5), lineWidth: 3)
                    )

                    // Sector label
                    let mid = (start + end) / 2 * .pi / 180
                    let lx = center.x + cos(mid) * radius * 0.6
                    let ly = center.y + sin(mid) * radius * 0.6
                    Text(sectors[i].label)
                        .font(.system(size: 18, weight: .bold))
                        .foregroundColor(.white)
                        .multilineTextAlignment(.center)
                        .position(x: lx, y: ly)
                }

                // Centre hub
                Circle()
                    .fill(GameColors.submitBg)
                    .frame(width: radius * 0.28, height: radius * 0.28)
                    .position(center)
                    .overlay(
                        Circle()
                            .stroke(Color.white, lineWidth: 3)
                            .frame(width: radius * 0.28, height: radius * 0.28)
                            .position(center)
                    )
            }
        }
    }
}

private struct SectorShape: Shape {
    let center: CGPoint
    let radius: CGFloat
    let startDeg: Double
    let endDeg: Double

    func path(in rect: CGRect) -> Path {
        var p = Path()
        p.move(to: center)
        p.addArc(
            center: center,
            radius: radius,
            startAngle: .degrees(startDeg),
            endAngle: .degrees(endDeg),
            clockwise: false,
        )
        p.closeSubpath()
        return p
    }
}

private struct Triangle: Shape {
    func path(in rect: CGRect) -> Path {
        var p = Path()
        p.move(to: CGPoint(x: rect.midX, y: rect.maxY))
        p.addLine(to: CGPoint(x: rect.minX + rect.width * 0.18, y: rect.minY))
        p.addLine(to: CGPoint(x: rect.maxX - rect.width * 0.18, y: rect.minY))
        p.closeSubpath()
        return p
    }
}

File: android-native/app/src/main/kotlin/com/wordwheel/game/ui/SpinWheel.kt
package com.wordwheel.game.ui

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.LinearOutSlowInEasing
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.drawText
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wordwheel.game.theme.GameColors
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin
import kotlin.random.Random

/**
 * One sector of the lottery wheel.
 *
 * The drawn sector size is uniform (360° / N); [weight] only controls
 * how often the spin lands on that sector.
 */
data class SpinSector(
    val coins: Int,
    val hints: Int,
    val color: Color,
    val weight: Int,
) {
    val label: String = when {
        coins > 0 && hints > 0 -> "+$coins\n+$hints💡"
        coins > 0 -> "+$coins"
        hints > 0 -> "+$hints💡"
        else -> "—"
    }
}

/** The default 8-sector wheel. Casual-friendly: small wins are common,
 *  rarer 50-coin or 3-hint jackpots keep the spin exciting. */
val DEFAULT_SECTORS: List<SpinSector> = listOf(
    SpinSector(coins = 5,  hints = 0, color = Color(0xFF50A0E6), weight = 28),
    SpinSector(coins = 10, hints = 0, color = Color(0xFFFFB400), weight = 22),
    SpinSector(coins = 15, hints = 0, color = Color(0xFF50A0E6), weight = 16),
    SpinSector(coins = 0,  hints = 1, color = Color(0xFF32C850), weight = 12),
    SpinSector(coins = 20, hints = 0, color = Color(0xFFFFB400), weight = 10),
    SpinSector(coins = 25, hints = 0, color = Color(0xFF50A0E6), weight = 6),
    SpinSector(coins = 50, hints = 0, color = Color(0xFFE65050), weight = 4),
    SpinSector(coins = 0,  hints = 3, color = Color(0xFF32C850), weight = 2),
)

/**
 * Modal lottery-wheel screen. Tap "SPIN" once → wheel animates to a
 * weighted-random sector → reward is reported via [onSpinResult].
 *
 * After the spin, tapping the dimmed background or the result button
 * dismisses the dialog.
 */
@Composable
fun SpinWheelDialog(
    sectors: List<SpinSector> = DEFAULT_SECTORS,
    onSpinResult: (SpinSector) -> Unit,
    onDismiss: () -> Unit,
) {
    var hasSpun by rememberSaveable { mutableStateOf(false) }
    var resultIndex by rememberSaveable { mutableIntStateOf(-1) }
    // Pair<finalRotationDegrees, sectorIndex> — set once the user taps
    // SPIN; observed by the LaunchedEffect below.
    var rotationTarget by remember { mutableStateOf<Pair<Float, Int>?>(null) }
    val rotation = remember { Animatable(0f) }

    val measurer = rememberTextMeasurer()

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(Color(0xC8000000))
            .clickable(enabled = hasSpun, onClick = onDismiss),
        contentAlignment = Alignment.Center,
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            modifier = Modifier
                .fillMaxWidth(0.86f)
                .shadow(elevation = 18.dp, shape = RoundedCornerShape(20.dp))
                .clip(RoundedCornerShape(20.dp))
                .background(GameColors.CompleteBg)
                .padding(20.dp),
        ) {
            Text(
                text = if (hasSpun) "Nice!" else "Daily Spin",
                color = Color.White,
                fontSize = 26.sp,
                fontWeight = FontWeight.Bold,
            )
            Spacer(Modifier.height(4.dp))
            Text(
                text = if (hasSpun)
                    sectors.getOrNull(resultIndex)?.let { sec ->
                        when {
                            sec.coins > 0 && sec.hints > 0 ->
                                "+${sec.coins} coins and +${sec.hints} hints"
                            sec.coins > 0 -> "+${sec.coins} coins"
                            sec.hints > 0 -> "+${sec.hints} hint${if (sec.hints > 1) "s" else ""}"
                            else -> ""
                        }
                    } ?: ""
                else "Tap below to spin — once per day",
                color = GameColors.StarYellow,
                fontSize = 14.sp,
            )
            Spacer(Modifier.height(20.dp))

            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .aspectRatio(1f),
                contentAlignment = Alignment.Center,
            ) {
                // Wheel
                Canvas(
                    modifier = Modifier
                        .fillMaxSize(0.92f)
                        .rotate(rotation.value),
                ) {
                    val radius = size.minDimension / 2f * 0.96f
                    val center = Offset(size.width / 2f, size.height / 2f)
                    val sweep = 360f / sectors.size

                    for ((i, sector) in sectors.withIndex()) {
                        // Sector centred at the top when its midpoint angle
                        // is -90°, so start = i*sweep − 90° − sweep/2.
                        val start = i * sweep - 90f - sweep / 2f
                        val path = Path().apply {
                            moveTo(center.x, center.y)
                            arcTo(
                                rect = androidx.compose.ui.geometry.Rect(
                                    center.x - radius, center.y - radius,
                                    center.x + radius, center.y + radius,
                                ),
                                startAngleDegrees = start,
                                sweepAngleDegrees = sweep,
                                forceMoveTo = false,
                            )
                            lineTo(center.x, center.y)
                            close()
                        }
                        drawPath(path, sector.color)
                        drawPath(
                            path,
                            color = Color(0xFFFFFFFF).copy(alpha = 0.5f),
                            style = Stroke(width = 3f),
                        )

                        val midAngleRad = (start + sweep / 2f) * PI.toFloat() / 180f
                        val labelDist = radius * 0.6f
                        val lx = center.x + cos(midAngleRad) * labelDist
                        val ly = center.y + sin(midAngleRad) * labelDist
                        val style = TextStyle(
                            color = Color.White,
                            fontSize = 18.sp,
                            fontWeight = FontWeight.Bold,
                        )
                        val layout = measurer.measure(sector.label, style)
                        drawText(
                            textLayoutResult = layout,
                            topLeft = Offset(
                                lx - layout.size.width / 2f,
                                ly - layout.size.height / 2f,
                            ),
                        )
                    }

                    drawCircle(
                        color = GameColors.SubmitBg,
                        radius = radius * 0.14f,
                        center = center,
                    )
                    drawCircle(
                        color = Color.White,
                        radius = radius * 0.14f,
                        center = center,
                        style = Stroke(width = 3f),
                    )
                }

                // Pointer (stays put; only the wheel rotates)
                Canvas(
                    modifier = Modifier
                        .align(Alignment.TopCenter)
                        .size(40.dp),
                ) {
                    val w = size.width
                    val h = size.height
                    val path = Path().apply {
                        moveTo(w / 2f, h)
                        lineTo(w * 0.18f, 0f)
                        lineTo(w * 0.82f, 0f)
                        close()
                    }
                    drawPath(path, Color(0xFFFFDC50))
                    drawPath(path, Color(0xFF000000).copy(alpha = 0.4f),
                             style = Stroke(width = 3f))
                }
            }

            Spacer(Modifier.height(20.dp))

            // SPIN button (disabled after the first spin)
            Box(
                modifier = Modifier
                    .clip(RoundedCornerShape(28.dp))
                    .background(
                        if (hasSpun) Color(0xFF555555) else Color(0xFF32C850)
                    )
                    .clickable(enabled = !hasSpun) {
                        val totalWeight = sectors.sumOf { it.weight }
                        val pick = Random.nextInt(totalWeight)
                        var acc = 0
                        var idx = 0
                        for ((i, sector) in sectors.withIndex()) {
                            acc += sector.weight
                            if (pick < acc) { idx = i; break }
                        }
                        resultIndex = idx
                        hasSpun = true

                        val sweep = 360f / sectors.size
                        // Wheel rotates clockwise; sector i centred at i*sweep
                        // (relative to the pointer at top). 5 full turns + the
                        // negative offset positions sector idx under pointer.
                        rotationTarget = (360f * 5 - idx * sweep) to idx
                    }
                    .padding(horizontal = 36.dp, vertical = 14.dp),
            ) {
                Text(
                    text = if (hasSpun) "TAP TO CLOSE" else "SPIN",
                    color = Color.White,
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Bold,
                )
            }
        }

        // Drive the actual rotation animation, then report the reward
        // back to the caller (which persists it).
        rotationTarget?.let { (target, idx) ->
            LaunchedEffect(target, idx) {
                rotation.animateTo(
                    targetValue = target,
                    animationSpec = tween(
                        durationMillis = 3200,
                        easing = LinearOutSlowInEasing,
                    ),
                )
                onSpinResult(sectors[idx])
            }
        }
    }
}
